### PR TITLE
[auth] Convert PKCS8 to PKCS1 private keys (for macOS)

### PR DESCRIPTION
**Pull request by dakcarto**
## Description
This addresses an issue I found during dev meeting in Essen (unreported until now).

Extracts the PrivateKey ASN.1 element of a DER-encoded PKCS#8 private key:

On some platforms, e.g. macOS, where the default SSL backend is not OpenSSL, a QSslKey can not be created using PKCS#8-formatted data. However, PKCS#8 private key ASN.1 structures contain the key data inside a wrapper describing the algorithm used, e.g. RSA, DSA, ECC etc. Extracted PrivateKey ASN.1 data can be used to create a compatible QSslKey, e.g. 'traditional' SSLeay RSA-specific PKCS#1.

By default OpenSSL 1.0.0+ returns private keys as PKCS#8, previously it was PKCS#1.

NOTE: This function requires ['libtasn1'](https://www.gnu.org/software/libtasn1/) development files and library, which is used to parse and extract the PrivateKey element from an ASN.1 PKCS#8 structure.

Also, adds more PEM- and DER-formatted client keys for testing.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: src/core/auth/qgsauthcertutils.cpp [hunk 1]
```diff
@@ -94,30 +101,34 @@ QMap<QString, QList<QgsAuthConfigSslServer> > QgsAuthCertUtils::sslConfigsGroupe
...
   }
+  file.close();
+
   return data;
 }
 
 QList<QSslCertificate> QgsAuthCertUtils::certsFromFile( const QString &certspath )
 {
   QList<QSslCertificate> certs;
   bool pem = certspath.endsWith( QLatin1String( ".pem" ), Qt::CaseInsensitive );
```
Comment (elpaso) [inline src/core/auth/qgsauthcertutils.cpp]: Not in scope with this PR, but here is one of the places were we should be more robust: if the file extension is not a reliable sign of the encoding, we should be trying both encoding before we give up, ignoring the file extension completely.

File: src/core/auth/qgsauthcertutils.cpp [hunk 2]
```diff
@@ -262,37 +273,249 @@ QStringList QgsAuthCertUtils::certKeyBundleToPem( const QString &certpath,
   return QStringList() << certpem << keypem << algtype;
 }
 
+bool QgsAuthCertUtils::pemIsPkcs8( const QString &keyPemTxt )
```
Comment (elpaso) [inline src/core/auth/qgsauthcertutils.cpp]: See my previous comment: can this function be used to detect PEM vs DER  instead of the file extension ?

**Review by elpaso — APPROVED** [review]
To be honest I didn't check all the key manipulation code, I'd need to study for quite some time in order to understand what it does with those different PKCS formats.
BTW if you really need an in-depth review of the key manipulation code, please let me know and I'll schedule the task.

File: src/core/auth/qgsauthcertutils.cpp (hunk 2 above)
Comment (dakcarto) [reply src/core/auth/qgsauthcertutils.cpp]: It doesn't do encoding determination. It specifically has to work with PEM in order to simply text search to determine PKCS format (much more involved if using DER). I suppose the signature could be more descriptive, e.g. `privateKeyPemIsPkcs8 (...)`.

File: src/core/auth/qgsauthcertutils.cpp (hunk 1 above)
Comment (dakcarto) [reply src/core/auth/qgsauthcertutils.cpp]: Yes, it needs addressed, but it is a larger refactoring, across many auth classes. Note this work already has an internal Boundless JIRA ticket for exactly that auth API change/refactoring.

**Comment by dakcarto** [comment]
@elpaso thanks for the code review. However, I realize it is not so easy to functionally test, as it requires macOS and the `libtasn1` extra library.

File: src/core/auth/qgsauthcertutils.cpp [hunk 3]
```diff
@@ -262,37 +273,249 @@ QStringList QgsAuthCertUtils::certKeyBundleToPem( const QString &certpath,
...
+  if ( oct_len == 0 )
+  {
+    QgsDebugMsg( QStringLiteral( "ERROR, asn1 privateKey value octet string empty" ) );
+    goto PKCS1DONE;
+  }
+
+  pkcs1 = QByteArray( oct_data, oct_len );
+
+  // !!! SENSITIVE DATA !!!
+  QgsDebugMsgLevel( QStringLiteral( "privateKey octet data as PEM: %1" ).arg( QString( pkcs1.toBase64() ) ), 9 );
```
Comment (dakcarto) [inline src/core/auth/qgsauthcertutils.cpp]: @elpaso, @m-kuhn, @nyalldawson, (and anyone else) thoughts on whether it is OK to leave this debug output active, but set the `QgsDebugMsgLevel` high, e.g. `9` here. Previously, I've left these kind of lines in the code, but commented out (with a warning not to leave uncommented). I can envision a circumstance where the env var `QGIS_DEBUG` might be manipulated somehow, thus exposing the credentials.

Comment (nyalldawson) [reply src/core/auth/qgsauthcertutils.cpp]: I'd hide it - " there's no reward in the risk" of exposing this to the log.



(+1million if anyone gets that reference)

**Comment by dakcarto** [comment]
Merged as the latest indention error makes no sense. Previous commit CI passed. Most recent commit would cause no issues.

**Pull request merged**